### ISSUE TITLE
NG: access image via immutable identifier

### ISSUE THREAD

**Issue by ncdc** [docker/docker-registry#804]
In OpenShift, we would like to be able to access an image (i.e. `docker pull/create/run`) via an immutable identifier that uniquely identifies an image for all points in time. Here are some use cases:

1. During a deployment of a group of containers, the same image must be used when creating each container. If the image we're using is `foo/bar:latest` (let's call this Rev1) and someone pushes an updated `foo/bar:latest` manifest in the middle of the deployment (Rev2), some containers might be created using Rev1 and others with Rev2. The correct behavior is for all containers to be running based on Rev1.
2. A group of containers has been deployed (`foo/bar:latest` - Rev1). As development continues, `foo/bar:latest` is updated several times, but none of these newer image manifests has been deployed yet. The user decides to scale up the existing deployment, so OpenShift needs to create and start new containers using the same image that is currently deployed (Rev1). `foo/bar:latest` can't be used because that's no longer Rev1 - it's now something else (e.g. Rev17).

With the v1 registry, we created a custom extension that responds to the `tag_created` signal, creates a new tag whose name is the image's id (since a v1 image has an id), and then http posts a payload to OpenShift with

1. namespace
2. repository
3. tag
4. image id
5. image metadata

OpenShift users can create deployment triggers that watch for changes to `foo/bar:latest` and then perform new deployments. When deploying, OpenShift inspects the image to find its id and then translates `foo/bar:latest` to `foo/bar:$image_id`. This, combined with our custom v1 registry extension, allows us to pull an image by its id. It also lets us deploy a specific image by id, as our deployments don't refer to `:latest` but instead to the id.

@dmp42 suggested that OpenShift could pull `foo/bar:latest`, generate a new tag (based on commit id, date/time, etc) that is unique and consistent for all time, push the new tag, and then use that when deploying. This creates a couple of problems:

1. The tag isn't immutable, which could be problematic if someone accidentally pushed an updated image manifest for a tag that is meant to never change
2. The new manifest would be signed by OpenShift, not by the user creating the image; if a user wants to run an image that he/she signed, we lose this ability.

It would really be nice to have immutable identifiers for image manifests that are consistent all the time.

@dmp42 @stevvooe @smarterclayton @wking thoughts?

**Comment by ncdc** [comment]
@stevvooe I definitely like your suggestion of having the registry compute the digest and return it to the client.

I know you had previously expressed concerns about the ability to perform GC if the registry retains copies of every revision of every manifest (unless they're somehow deleted, either manually by a user or automatically via some sort of policy). Are these concerns still an issue for you, or are they mitigated by the delete mechanics listed in bullet 4 above?

cc @smarterclayton - any additional thoughts on the previous comment?

**Comment by stevvooe** [comment]
@ncdc

> Are these concerns still an issue for you, or are they mitigated by the delete mechanics listed in bullet 4 above?

They are mitigated by bullet 4 above. This approach saves everything unless specifically asked to delete it. An external webhook service can then be used to control manifest lifecycle. This keeps GC simple (ref counting) and separates it from lifecycle management. It also reduces the possibility of data loss upon manifest updates.

**Comment by wking** [comment]
On Tue, Dec 30, 2014 at 11:50:48AM -0800, Stephen Day wrote:
> 2. The "digest" of the manifest is the sha256 of the "unsigned"
>    portion of manifest, with sorted object keys. This should only be
>    calculated by the registry for the time being.
> 3. `PUT` operations on the manifest are no longer destructive. If
>    the content is different, the "tag" is updated to point at the
>    new content. All revisions remain addressable by digest.

The PUT operation will still be descructive if the unsigned portion of
the manifest changes (e.g. new signatures or signers).  I'm fine with
that, but thought I'd point it out for completeness.

**Comment by stevvooe** [comment]
@wking We may lean towards just taking the entire hash of the content to address this. There seem to be problems with specialized hashes no matter what way we try to cut this up. We may want to discuss storing the signatures separately from the manifest.

**Comment by stevvooe** [comment]
@wking We may actually be able to merge the signatures on the registry side.

**Comment by wking** [comment]
On Mon, Jan 05, 2015 at 05:10:31PM -0800, Stephen Day wrote:
> There seem to be problems with specialized hashes no matter what way
> we try to cut this up.

And the way to avoid this is to just hash the whole thing ;).  If “the
whole thing” is too much (and I think it is), I'd break the manifest
up into a bunch of individually content-addressable objects (e.g. the
layer tarballs, the image metadata, signed name/tag/image
certifications).

> We may want to discuss storing the signatures separately from the
> manifest.

Sounds good to me (this is where I started out with docker/docker#6070
and [1]).  I suggested (but didn't elaborate on) additional endpoints
for independently distributing opaque signatures too [2].  Of course,
those proposals were framed in terms of my preferred thin registry,
with all of the validation happening on the client side.  With the
current thick registry, you'd have to specify a signature format as
well as signature endpoints and have the registry validate signatures
as they're uploaded, but I think all of that has been worked out for
the new registry code anyway.

[1]: https://github.com/docker/docker/issues/8093#issuecomment-56956721
[2]: https://github.com/docker/docker/pull/6070#issuecomment-56952873
